### PR TITLE
initial revision of external module augmentations

### PR DESCRIPTION
**Pull request by vladima**
Module augmentation is a declaration of ambient module that directly nested either in external module or in top level ambient external module. 
Name of module augmentation is resolved using the same set of rules as module specifiers in import \ export declarations. 
If name is successfully resolved to some external module then declarations in module augmentation are merged with declarations inside module  using standard rules for declaration merging. 
Module augmentations cannot add new items to the top level scope but rather patch existing declarations.

for example
```ts
// observable.ts
export class Observable<T> {}
```
```ts
// map.ts
import {Observable} from "./observable";
Observable.prototype.map = function () { /* ... */ }

declare module "./observable" {
    interface Observable<T> {
        map<U>(proj: (el: T) => U): Observable<U>;
    }
}
```
```t
// consumer.ts
import {Observable} from "./observable";
import "./map";
let o: Observable<number>;
o.map(x => x.toFixed());
```

Here module `map` can declare that internally it will patch `Observable` type and add `map` function to it.

Fixes: #5269,  #6022

Pending work:

- [x] - add syntactic support for augmenting global scope. Currently augmentation for the global scope is defined as ambient module with special name `"/"` but this is just a placeholder. The version that we ended up during the design meeting was:
```ts
// in external module
export {};
declare global {}
```
or 
```ts
// in script file
declare module "array" {
    global {}
}
```

### CODE REVIEW

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -109,6 +109,7 @@ namespace ts {
         let blockScopeContainer: Node;
         let lastContainer: Node;
         let seenThisKeyword: boolean;
+        let isSourceFileExternalModule: boolean;
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: I don't see this getting used anywhere apart from the later assignment.

File: src/compiler/binder.ts [hunk 2]
```diff
@@ -348,7 +350,12 @@ namespace ts {
                 //   2. When we checkIdentifier in the checker, we set its resolved symbol to the local symbol,
                 //      but return the export symbol (by calling getExportSymbolOfValueSymbolIfExported). That way
                 //      when the emitter comes back to it, it knows not to qualify the name if it was found in a containing scope.
-                if (hasExportModifier || container.flags & NodeFlags.ExportContext) {
+
+                // NOTE: Nested ambient modules always should go to to 'locals' table to prevent their automatic merge
+                //       during global merging in the checker. Why? The only case when ambient module is permitted inside another module is module augmentation 
+                //       and this case is specially handled. Module augmentations should only be merged with original module definition 
+                //       and should never be merged directly with other augmentation and the latter case would be possible is automatic merge is allowed.
```
Comment (DanielRosenwasser) [inline src/compiler/binder.ts]: add a comma after "augmentation" so this reads more easily, and add I think you meant "if" instead of "is" on this line.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -368,6 +368,32 @@ namespace ts {
...
+            if (isNameOfGlobalAugmentation(moduleName)) {
+                mergeSymbolTable(globals, moduleAugmentation.symbol.exports);
+            }
+            else {
+                // find a module that about to be augmented 
+                let mainModule = resolveExternalModuleNameWorker(moduleName, moduleName, Diagnostics.Invalid_module_name_in_augmentation_module_0_cannot_be_found);
+                if (!mainModule) {
+                    return;
+                }
+                // is module symbol is already merged - it is safe to use it.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: The module symbol

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: If the module symbol has already been merged

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -14176,12 +14302,16 @@ namespace ts {
...
-                // Relative external module names are not permitted.
-                error(node, Diagnostics.Import_or_export_declaration_in_an_ambient_module_declaration_cannot_reference_module_through_relative_module_name);
-                return false;
+                // we have already reported errors on top level imports\exports in external module augmentations in checkModuleDeclaration
+                // no need to do this again.
+                if (!isTopLevelInExternalModuleAugmentation(node)) {
+                    // TypeScript 1.0 spec (April 2013): 12.1.6
+                    // An ExternalImportDeclaration in an AmbientExternalModuleDeclaration may reference
+                    // other external modules only through top - level external module names.
+                    // Relative external module names are not permitted.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This comment needs to be updated.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -14176,12 +14302,16 @@ namespace ts {
...
+                    error(node, Diagnostics.Import_or_export_declaration_in_an_ambient_module_declaration_cannot_reference_module_through_relative_module_name);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This should be `An import or export declaration in an ambient module declaration cannot reference modules with relative path names.`

File: src/compiler/binder.ts (hunk 1 above)
Comment (vladima) [inline src/compiler/binder.ts]: right, forgot to remove that. Thanks

File: src/compiler/binder.ts (hunk 2 above)
Comment (vladima) [inline src/compiler/binder.ts]: done

File: src/compiler/checker.ts (hunk 3 above)
Comment (vladima) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 4 above)
Comment (vladima) [inline src/compiler/checker.ts]: why? it still describes the intended behavior for ambient modules

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -14133,19 +14178,100 @@ namespace ts {
...
-                        error(node.name, Diagnostics.Ambient_modules_cannot_be_nested_in_other_modules_or_namespaces);
+                    if (isExternalModuleAugmentation(node)) {
+                        // if symbol of augmentation is not merged this means that either
+                        // - this is an augmentation of the global scope 
+                        // or
+                        // - this augmentation was not merged with main definition of the module
+                        //   error should already be reported so all errors in the body of augmentation can be ignored.
+                        const checkBody = isNameOfGlobalAugmentation(<LiteralExpression>node.name) || (getSymbolOfNode(node).flags & SymbolFlags.Merged);
+                        if (checkBody) {
+                            const globalAugmentation = isNameOfGlobalAugmentation(<LiteralExpression>node.name);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: isGlobalAugmentation

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -14133,19 +14178,100 @@ namespace ts {
...
-                // Checks for ambient external modules.
                 if (isAmbientExternalModule) {
-                    if (!isGlobalSourceFile(node.parent)) {
-                        error(node.name, Diagnostics.Ambient_modules_cannot_be_nested_in_other_modules_or_namespaces);
+                    if (isExternalModuleAugmentation(node)) {
+                        // if symbol of augmentation is not merged this means that either
+                        // - this is an augmentation of the global scope 
+                        // or
+                        // - this augmentation was not merged with main definition of the module
+                        //   error should already be reported so all errors in the body of augmentation can be ignored.
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I don't actually understand this error case.

File: src/compiler/checker.ts [hunk 8]
```diff
@@ -14133,19 +14178,100 @@ namespace ts {
                     }
                 }
 
-                // Checks for ambient external modules.
                 if (isAmbientExternalModule) {
-                    if (!isGlobalSourceFile(node.parent)) {
-                        error(node.name, Diagnostics.Ambient_modules_cannot_be_nested_in_other_modules_or_namespaces);
+                    if (isExternalModuleAugmentation(node)) {
+                        // if symbol of augmentation is not merged this means that either
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Maybe phrase this "An augmentation symbol should not be merged if:"

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Can you explain what code path you're documenting in case I misunderstood?

File: src/compiler/checker.ts [hunk 9]
```diff
@@ -14133,19 +14178,100 @@ namespace ts {
...
+                            for (const statement of (<ModuleBlock>node.body).statements) {
+                                checkBodyOfModuleAugmentation(statement, globalAugmentation);
+                            }
+                        }
                     }
-                    if (isExternalModuleNameRelative(node.name.text)) {
-                        error(node.name, Diagnostics.Ambient_module_declaration_cannot_specify_relative_module_name);
+                    else if (isGlobalSourceFile(node.parent)) {
+                        if (isExternalModuleNameRelative(node.name.text)) {
+                            error(node.name, Diagnostics.Ambient_module_declaration_cannot_specify_relative_module_name);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This should be `An ambient module declaration cannot specify a relative module name.`

File: src/compiler/checker.ts [hunk 10]
```diff
@@ -14133,19 +14178,100 @@ namespace ts {
...
+                                checkBodyOfModuleAugmentation(el, isGlobalAugmentation);
+                            }
+                        }
+                        else {
+                            checkBodyOfModuleAugmentation(decl, isGlobalAugmentation);
+                        }
+                    }
+                    break;
+                case SyntaxKind.ExportDeclaration:
+                    grammarErrorOnFirstToken(node, Diagnostics.Exports_are_not_permitted_in_module_augmentations);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: This is a little misleading. Maybe the message should be "Exports must occur at declaration sites in a module augmentation."

Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Never mind, they're just not allowed period it sounds like.

File: src/compiler/checker.ts [hunk 11]
```diff
@@ -15473,13 +15603,27 @@ namespace ts {
                 bindSourceFile(file, compilerOptions);
             });
 
+            let mergeAugmentations = false;
             // Initialize global symbol table
             forEach(host.getSourceFiles(), file => {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Pretty sure this can just be turned into a for-of

File: src/compiler/checker.ts (hunk 7 above)
Comment (vladima) [inline src/compiler/checker.ts]: you have augmentation for some module './m' but actual definition of the module cannot be found. We've already reported an error that './m' cannot be found so there is no need to dive into the body of augmentation at for every declaration say that it is incorrect because it introduces a new name in module

File: src/compiler/checker.ts [hunk 12]
```diff
@@ -15473,13 +15603,27 @@ namespace ts {
...
                     mergeSymbolTable(globals, file.locals);
                 }
+                mergeAugmentations = mergeAugmentations || file.moduleAugmentations.length > 0;
             });
 
+            if (mergeAugmentations) {
+                // merge module augmentations.
+                // this needs to be done after global symbol table is initialized to make sure that all ambient modules are indexed 
+                for (const file of host.getSourceFiles()) {
+                    if (file.moduleAugmentations.length) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Why do you need to check the length? The for-of will take care of that anyway.

File: src/compiler/checker.ts [hunk 13]
```diff
@@ -15473,13 +15603,27 @@ namespace ts {
...
+                        for (const augmentation of file.moduleAugmentations) {
+                            mergeModuleAugmentation(augmentation);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: You could make this `forEach(file.moduleAugmentations, mergeModuleAugmentation)

Comment (vladima) [inline src/compiler/checker.ts]: I don't see large benefit of using `forEach` over `for..of` here.

File: src/compiler/checker.ts (hunk 11 above)
Comment (vladima) [inline src/compiler/checker.ts]: it can and I think we should make a sweep over the compiler to convert `forEach` to `for..of` at some point. But this should be a separate PR.

File: src/compiler/checker.ts (hunk 12 above)
Comment (vladima) [inline src/compiler/checker.ts]: true, will change

File: src/compiler/program.ts [hunk 14]
```diff
@@ -887,59 +914,75 @@ namespace ts {
...
+                        if (!inAmbientModule || !isExternalModuleNameRelative((<LiteralExpression>moduleNameExpr).text)) {
+                            (imports || (imports = [])).push(<LiteralExpression>moduleNameExpr);
+                        }
+                        break;
+                    case SyntaxKind.ModuleDeclaration:
+                        if ((<ModuleDeclaration>node).name.kind === SyntaxKind.StringLiteral && (inAmbientModule || node.flags & NodeFlags.Ambient || isDeclarationFile(file))) {
+                            const moduleName = <LiteralExpression>(<ModuleDeclaration>node).name;
+                            // Ambient module declarations can be interpreted as augmentations for some existing external modules.
+                            // This will happen in two cases:
+                            // - if current file is external module then module augmentation is a ambient module declaration defined in the top level scope
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: I think you mean "if the ambient module declaration is in the top level scope of an external module file, then it is a module augmentation"

File: src/compiler/program.ts [hunk 15]
```diff
@@ -887,59 +914,75 @@ namespace ts {
...
+                            // - if current file is not external module then module augmentation is an ambient module declaration with non-relative module name
```
Comment (DanielRosenwasser) [inline src/compiler/program.ts]: "if the current file is not an external module (e.g. an ambient declaration file), then if the module name is non-relative, the declaration is a module augmentation"

File: tests/baselines/reference/moduleAugmentationCollidingNamesInAugmentation1.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
+//// [tests/cases/compiler/moduleAugmentationCollidingNamesInAugmentation1.ts] ////
+
+//// [map1.ts]
+
+import { Observable } from "./observable"
+
+(<any>Observable.prototype).map = function() { }
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/moduleAugmentationCollidingNamesInAugmentation1.js]: Do you need to cast to `any` here? I would think that the type of `prototype` would be augmented by the interface correctly.

File: tests/baselines/reference/moduleAugmentationDeclarationEmit1.js [hunk 17]
```diff
@@ -0,0 +1,68 @@
+//// [tests/cases/compiler/moduleAugmentationDeclarationEmit1.ts] ////
+
+//// [map.ts]
+
+import { Observable } from "./observable"
+
+(<any>Observable.prototype).map = function() { }
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: Do you have to assert this to `any`? Why doesn't `map` just get added to the prototype?

File: src/compiler/checker.ts (hunk 13 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: Well you have a function that can act as a simple callback, so it seems unnecessary to explicitly write the loop

File: src/compiler/checker.ts (hunk 11 above)
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: I did that and I don't remember why we never got it pulled in (#2428)

File: tests/baselines/reference/moduleAugmentationDeclarationEmit1.js (hunk 17 above)
Comment (vladima) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: because type of prototype is `Observable<any>` and it does not have `map` property

**Comment by DanielRosenwasser** [comment]
What exactly is the recommended way to augment a default export?

**Comment by vladima** [comment]
it is not possible

File: tests/baselines/reference/moduleAugmentationDeclarationEmit1.js (hunk 17 above)
Comment (DanielRosenwasser) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: But you're declaring an extension to `Observable<T>` below, so doesn't that mean `Observable<any>` should have a member named `map` of type `map<U>(proj: (e: any) => U): Observable<U>`?

Comment (vladima) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: this is just an example that in module `map` somebody can patch the prototype of `Observable` and add new functionality to it and *also* describe what was done to our compiler via module augmentation. What was added to the prototype and how it was done is not really important here.

File: src/compiler/checker.ts (hunk 13 above)
Comment (vladima) [inline src/compiler/checker.ts]: it is easier to reason about perf implications when you have plain old loop

File: tests/baselines/reference/moduleAugmentationDeclarationEmit1.js (hunk 17 above)
Comment (DanielRosenwasser) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: I think to some degree it is, since users who have not patched up the prototype prior to this will attempt to do this without asserting to `any`. If we can, let's amend this test to reflect the intended real world usage, or add another test.

Comment (vladima) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: our test suite is not a set of examples and I'd prefer not to pollute it with things that are non-relevant to feature being tested. Once we finalize the design and check-in the implementation we can add as many examples to the handbook as we want

**Comment by DanielRosenwasser** [comment]
This also appears to take care of #2784. You might want to take note and close that issue when this goes in.

File: src/compiler/declarationEmitter.ts [hunk 18]
```diff
@@ -139,6 +141,14 @@ namespace ts {
                 allSourcesModuleElementDeclarationEmitInfo = allSourcesModuleElementDeclarationEmitInfo.concat(moduleElementDeclarationEmitInfo);
                 moduleElementDeclarationEmitInfo = [];
             }
+
+            if (!isBundledEmit && isExternalModule(sourceFile) && sourceFile.moduleAugmentations.length && !resultHasExternalModuleIndicator) {
```
Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: what about external modules with no exports or no module augmentation? soemthing like:
```ts
import {a} from "./mod";
// do something with a
```

Comment (mhegazy) [inline src/compiler/declarationEmitter.ts]: i thought we will do this for any external module, that we never emitted any import, export, or module augmentation for.

File: src/compiler/checker.ts [hunk 19]
```diff
@@ -14142,7 +14185,13 @@ namespace ts {
         function checkModuleDeclaration(node: ModuleDeclaration) {
             if (produceDiagnostics) {
                 // Grammar checking
-                const isAmbientExternalModule = node.name.kind === SyntaxKind.StringLiteral;
+                const isGlobalAugmentation = isGlobalScopeAugmentation(node);
+                const inAmbientContext = isInAmbientContext(node);
+                if (isGlobalAugmentation && !inAmbientContext) {
+                    error(node.name, Diagnostics.Augmentations_for_the_global_scope_should_have_declare_modifier_unless_they_appear_in_already_ambient_context);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: can we just say "Augmentations_for_the_global_scope_should_have_declare_modifier", we know they do not appear in one.

File: src/compiler/checker.ts [hunk 20]
```diff
@@ -14186,19 +14235,101 @@ namespace ts {
...
+                        }
                     }
                 }
             }
             checkSourceElement(node.body);
         }
 
+        function checkBodyOfModuleAugmentation(node: Node, isGlobalAugmentation: boolean): void {
+            switch (node.kind) {
+                case SyntaxKind.VariableStatement:
```
Comment (mhegazy) [inline src/compiler/checker.ts]: we should also add ExportAssignment.

Comment (vladima) [inline src/compiler/checker.ts]: done

**Comment by mhegazy** [comment]
:+1:

**Comment by vladima** [comment]
@ahejlsberg do you have any comments?

File: src/compiler/program.ts [hunk 21]
```diff
@@ -1083,14 +1126,28 @@ namespace ts {
...
+            if (file.imports.length || file.moduleAugmentations.length) {
                 file.resolvedModules = {};
-                const moduleNames = map(file.imports, name => name.text);
+                const moduleNames: string[] = [];
+                for (const name of file.imports) {
+                    moduleNames.push(name.text);
+                }
+                for (const name of file.moduleAugmentations) {
+                    moduleNames.push(name.text);
+                }
```
Comment (ahejlsberg) [inline src/compiler/program.ts]:
```typescript
const moduleNames = concatenate(file.imports, file.moduleAugmentations);
```

File: src/compiler/program.ts [hunk 22]
```diff
@@ -498,15 +515,25 @@ namespace ts {
...
 
                     if (resolveModuleNamesWorker) {
-                        const moduleNames = map(newSourceFile.imports, name => name.text);
+                        const moduleNames: string[] = [];
+                        for (const moduleName of newSourceFile.imports) {
+                            moduleNames.push(moduleName.text);
+                        }
+                        for (const moduleName of newSourceFile.moduleAugmentations) {
+                            moduleNames.push(moduleName.text);
+                        }
```
Comment (ahejlsberg) [inline src/compiler/program.ts]:
```typescript
const moduleNames = concatenate(newSourceFile.imports, newSourceFile.moduleAugmentations);
```

Comment (ahejlsberg) [inline src/compiler/program.ts]: Just FYI, the `concatenate` function treats `undefined` the same as an empty array and only allocates a new array if both arguments are defined. Otherwise it just returns the one defined argument. Since it'll be rare for both to be defined here, it's fine to use this function instead of the imperative loops.

File: src/compiler/checker.ts [hunk 23]
```diff
@@ -14186,19 +14235,102 @@ namespace ts {
...
-                // Checks for ambient external modules.
                 if (isAmbientExternalModule) {
-                    if (!isGlobalSourceFile(node.parent)) {
-                        error(node.name, Diagnostics.Ambient_modules_cannot_be_nested_in_other_modules_or_namespaces);
+                    if (isExternalModuleAugmentation(node)) {
+                        // body of the augmentation should be checked for consistency only if augmentation was applied to its target (either global scope or module)
+                        // otherwise we'll be swamped in cascading errors. 
+                        // We can detect if augmentation was applied using following rules:
+                        // - augmentation for a global scope is always applied
+                        // - augmentation for some external module is applied if symbol for augmentation is merged (it was combined with target module).
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: But just so I'm clear, this is because the module to augment wasn't found, and we'll have reported that error elsewhere?

File: src/compiler/checker.ts [hunk 24]
```diff
@@ -14186,19 +14235,102 @@ namespace ts {
...
+                            // This means that this is declaration of ambient module that is located in other module or namespace which is prohibited.
+                            error(node.name, Diagnostics.Ambient_modules_cannot_be_nested_in_other_modules_or_namespaces);
+                        }
                     }
                 }
             }
             checkSourceElement(node.body);
         }
 
+        function checkBodyOfModuleAugmentation(node: Node, isGlobalAugmentation: boolean): void {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: This should be `checkModuleAugmentationElement`. It checks and element in the body, not the body itself.

File: src/compiler/program.ts (hunk 22 above)
Comment (vladima) [inline src/compiler/program.ts]: ok

File: src/compiler/checker.ts (hunk 23 above)
Comment (vladima) [inline src/compiler/checker.ts]: yes, it is done [here](https://github.com/Microsoft/TypeScript/pull/6213/files#diff-c3ed224e4daa84352f7f1abcd23e8ccaR394)

File: src/compiler/checker.ts [hunk 25]
```diff
@@ -14186,19 +14235,102 @@ namespace ts {
...
+                    if ((<ImportEqualsDeclaration>node).moduleReference.kind !== SyntaxKind.StringLiteral) {
+                        error((<ImportEqualsDeclaration>node).name, Diagnostics.Module_augmentation_cannot_introduce_new_names_in_the_top_level_scope);
+                        break;
+                    }
+                    // fallthrough
+                case SyntaxKind.ImportDeclaration:
+                    grammarErrorOnFirstToken(node, Diagnostics.Imports_are_not_permitted_in_module_augmentations_Consider_moving_them_to_the_enclosing_external_module);
+                    break;
+                default:
+                    const symbol = getSymbolOfNode(node);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: You're assuming here that `getSymbolOfNode` can be called on *any* possible top-level, and that if that produces a symbol, the node is a declaration node. Not sure this is safe, particularly since you might see statements here (that would be flagged as errors elsewhere). I think the `default:` clause above should be replaced with a list of actual node kinds that you want to check.

File: src/compiler/checker.ts (hunk 24 above)
Comment (vladima) [inline src/compiler/checker.ts]: done

File: src/compiler/checker.ts (hunk 25 above)
Comment (vladima) [inline src/compiler/checker.ts]: what about checking `isDeclaration(node)` first?

Comment (ahejlsberg) [inline src/compiler/checker.ts]: I think you need the real list of nodes. It is a relatively short list no doubt, where `isDeclaration` has a whole bunch of node kinds you don't want.

Comment (vladima) [inline src/compiler/checker.ts]: ok, then if we need to check only top level items we can limit the list to
```ts
            case SyntaxKind.ClassDeclaration:
            case SyntaxKind.EnumDeclaration:
            case SyntaxKind.FunctionDeclaration:
            case SyntaxKind.InterfaceDeclaration:
            case SyntaxKind.ModuleDeclaration:
            case SyntaxKind.TypeAliasDeclaration:
```

**Comment by vladima** [comment]
build failure is related to #6478

**Comment by vladima** [comment]
@ahejlsberg done, do you have any other comments?

File: src/compiler/checker.ts [hunk 26]
```diff
@@ -14190,19 +14239,110 @@ namespace ts {
...
+                    if (symbol) {
+                        // module augmentations cannot introduce new names on the top level scope of the module
+                        // this is done it two steps
+                        // 1. quick check - if symbol for node is not merged - this is local symbol to this augmentation - report error
+                        // 2. main check - report error if value declaration of the parent symbol is module augmentation)
+                        let reportError = !(symbol.flags & SymbolFlags.Merged);
+                        if (!reportError) {
+                            if (isGlobalAugmentation) {
+                                // global symbol should not have parent since it is not explicitly exported
+                                reportError = symbol.parent !== undefined;
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: The error message below seems a bit odd for this case, but maybe it's fine.

Comment (vladima) [inline src/compiler/checker.ts]: for global case we can easily lift this restriction since it is possible to add new entries to global scope from within a module. Currently the fact that we have it is only because of consistency in behavior for augmentations.

**Comment by masaeedu** [comment]
In `map.ts`: since the ambient module declaration advertises to the compiler that `Observable<T>` has a `map` member, would you be able to attach a function of the appropriate kind to the prototype without casting to `<any>`?

File: src/compiler/checker.ts (hunk 26 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: No sure I understand, maybe we're talking about different issues. This code is saying that you get an error when you have an exported member in a global augmentation block, right? I'm thinking it should never be valid to use `export` in a global augmentation block.

File: src/compiler/checker.ts [hunk 27]
```diff
@@ -14226,19 +14275,110 @@ namespace ts {
...
+                            }
+                            else {
+                                // this symbol contains only merged content from external modules and augmentations so it should always be exported (parent !== undefined)
+                                // and parent should have value side (valueDeclaration !== undefined)
+                                Debug.assert(symbol.parent !== undefined && symbol.parent.valueDeclaration !== undefined);
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: I have a hard time reasoning about this.

File: src/compiler/checker.ts [hunk 28]
```diff
@@ -15568,13 +15714,25 @@ namespace ts {
...
                     mergeSymbolTable(globals, file.locals);
                 }
+                mergeAugmentations = mergeAugmentations || file.moduleAugmentations.length > 0;
             });
 
+            if (mergeAugmentations) {
+                // merge module augmentations.
+                // this needs to be done after global symbol table is initialized to make sure that all ambient modules are indexed 
+                for (const file of host.getSourceFiles()) {
+                    for (const augmentation of file.moduleAugmentations) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts]: Maybe just collect all augmentations into a single array and then `for..of` on that so you don't have to visit all the source files again. Very few source files are going to have augmentations. Code would be shorter too.

Comment (vladima) [inline src/compiler/checker.ts]: :white_check_mark:

File: src/compiler/checker.ts (hunk 26 above)
Comment (vladima) [inline src/compiler/checker.ts]: the idea of this check is following: `symbol.parent`  will be `undefined` if symbol was initially defined in the global scope. If `symbol.parent` is not `undefined` this means that this symbol was  declared inside augmentation and will declare new entry in the global scope which is disallowed in current design

File: src/compiler/checker.ts (hunk 27 above)
Comment (vladima) [inline src/compiler/checker.ts]: this assert is redundant. it basically says: augmentation can only change exported symbol (`parent !== undefeind`) that really has a declaration (not a virtual symbol like `prototype`) - this should always be the case. I'll remove it

File: src/compiler/checker.ts (hunk 26 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: > If symbol.parent is not undefined this means that this symbol was declared inside augmentation

Why is that so? `symbol.parent` is non-undefined when the symbol is exported, but it might still be declared inside the augmentation. I'm not getting this.

File: src/compiler/checker.ts (hunk 27 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: So did you mean `symbol.valueDeclaration !== undefined`? You currently have `symbol.parent.valueDeclaration`.

Comment (vladima) [inline src/compiler/checker.ts]: `symbol.parent.valueDeclaration` - is an augmentation: `declare module ...`. `symbol.valueDeclaration` is declaration inside the augmentation. Currently we verify: `thing that exports this symbol is module augmentation`

File: src/compiler/checker.ts (hunk 26 above)
Comment (vladima) [inline src/compiler/checker.ts]: symbols whose initial declaration is in global scope are never exported and `symbol.parent` for them is always `undefined`. if `symbol.parent !== undefined` this means that this symbol is exported from somewhere and as a consequence initially defined in augmentation not in global scope

**Comment by blesh** [comment]
Awesome! Great stuff. Thank you!

**Comment by DanielRosenwasser** [comment]
:tada:  :clap: :clap: :clap: :clap: :clap: :clap: :clap:

**Comment by weswigham** [comment]
:+1:

File: tests/baselines/reference/moduleAugmentationDeclarationEmit1.js (hunk 17 above)
Comment (mihailik) [inline tests/baselines/reference/moduleAugmentationDeclarationEmit1.js]: I would help if conversations like these bore their fruits as extensive comments in the code.

If the members of the core team interpret that stuff differently, imagine the confusion of an external observer. Consider a day when both of you have left the team and no one knows `<any>` reason to have that cast in there.
